Procedure

ldesousa opened the pull request:
Complete collections of classes and individuals for Observations, Properties and Procedures related with wet chemistry analysis. Also adds a new sub-ontology with units of measure specific to these new properties.

ldesousa commented:
I will send a Doodle around to set up a meeting to discuss this pull request.

ldesousa commented:
For reference, the latest version of [the list with properties and procedures](https://drive.google.com/file/d/1-UAEhozkPpbbYxDbKIF3xU6jFm_I4Y28/view?usp=sharing).

Montanaz0r commented:
Hello Luis, 

as we just discussed at the meeting, here is the list of pending issues:

1) Naming convention: Classes should be represented with the name starting
from a capital letter. Example: availableWaterHoldingCapacity →
AvailableWaterHoldingCapacity. The same goes for classes created to
represent values example: availableWaterHoldingCapacityValue →
AvailableWaterHoldingCapacityValue

**UPDATE:** **We all agreed this is a mistake and will be corrected with an update**

2) Using same property in different Observations, is it desirable? Example:
glosis_lh:aluminiumProperty used in glosis_lh:aluminiumExchangeablebases
and in glosis_lh:aluminiumExtractableelements

**UPDATE: We confirmed that this was indeed done on purpose and different Observations may relate to the same property.**

3) CamelCase naming convention is not always followed, example:
glosis_lh:aluminiumExchangeablebases should be
glosis_lh:aluminiumExchangeableBases

**UPDATE: We all agreed we should follow the CamelCase convention. This will be adjusted with an update to the PR.**

4) glosis_procedure module is whole shifted by one procedure.

**UPDATE: We agreed that this is an issue and can be easily fixed by tweaking the automation script that Luis made.**

5) In our original glosis_lh ontology we got two classes:
glosis_lh:MoistConsistency and glosis_lh:ConsistenceMoist which seems to be
referencing the same thing, can we confirm that somehow?

**UPDATE: This is something that has to be discussed/solved in a wider group.** 

6) Additions to glosis_lh are summarised in the spreadsheet.

https://docs.google.com/spreadsheets/d/1QwKwO8pWnfoWNVMxgS6889Xzq_wczoAfYwODE3zGDpc/edit?usp=sharing

**UPDATE:**

**Here is the link to the spreadsheet that comments on all 61 classes that were added to the glosis_lh in the latest PR. Some of them are shadowing/extending the code list we had. If we indeed agree on extending those, we should decide if we remove the corresponding code list.**

**Others are possibly a duplicated classes, which were already present in the glosis_lh module. If that is the case, and a duplicate was confirmed, we should stick with updating the existing Observation with Procedure, unit, and so on.**

**Some of them seem to be completely NEW, but again there is some concern about the overlap with memberPropertyValue codelist.**

**In two cases there is a missing prefix (this is what we deduced based on the naming convention). Luis will check the script, perhaps there was some white space and the script has not included the proper prefix.**

Montanaz0r commented:
Luis, please let me know in case you got any questions or issues with accessing the spreadsheet.

Montanaz0r commented:
regarding the Python scripts you mentioned, I would suggest to up them under **glosis/utils/** unless @rapw3k got a better idea

ldesousa commented:
@Montanaz0r a new iteration is ready for review. I think the Camel case issues are all addressed, but please verify.

Regarding duplicates, here is a summary of what I did:

- `glosis_lh:bulkDensity` - the new class is not correct. Retaining the two existing classes and updating with relations to Procedures.

- `glosis_lh:coarseFragments` - looks like fragments and coarse fragments are different things, with different units, leaving as it is for now. This will need to be revisited by soil scientists.

- `glosis_lh:electricalConductivity` - actual duplicate, updated existing class(es).

- `glosis_lh:hydraulicConductivity` - actual duplicate, updated existing class(es). 

- `glosis_lh:organicMatter` - can't find the duplicate, discuss.

- `glosis_lh:porosity` - actual duplicate, updated existing class(es). 

- `glosis_lh:sandTexturesasicl`, `glosis_lh:siltTexturesasicl` and `glosis_lh:clayTexturesasicl`: the existing properties refer to texture classes, whereas the new are texture fractions. I added the suffix "Fraction" to differentiate.

Montanaz0r commented:
Hello Luis,

Thank you for submitting the changes! As far as my initial list is concerned, most of it is already solved thanks to your update:

1. The naming convention is solved and classes are starting from the upper case letter now,
2. Already discussed and solved before,
3. All additions are now following CamelCase convention; solved,
4. The procedure module is no longer shifted, now everything looks fine. Solved,
5. No update on this one, we need to discuss is it at the next meeting,
6. Regarding issues that were summarised in the excel sheet, here is what still needs some verification/work/comment:

 - **glosis_lh:bulkDensity** was deleted and instead of that we extended **BulkDensityFineEarth** and **BulkDensityWholeSoil**. Each of them got their procedure. How about two other BulkDensity classes that can be encountered in the glosis layer-horizon ontology, namely: **glosis_lh:BulkDensityPea**t and **glosis_lh:BulkDensityMineral**. Are we able to apply some kind of procedure to those two as well, or they should be treated separately from the two classes that you extended?

- **glosis_lh:coarseFragments** still needs an evaluation from soil scientists but until that happens we will treat them as different things considering the difference in the unit,

- **glosis_lh:cecSoil** this one seems to be missing now, it was marked as a completely new class in my spreadsheet, but now it is missing. What has happened?

- **glosis_lh:aluminiumExtractableelements** and **glosis_lh:aluminiumTotalelements** are now having a typo, they both are named as aluminum instead of alumin**i**um!

- we are suspecting that **glosis_lh:organicMatter** is a duplicate of **glosis_cm:OrganicMatterClass**. Please note this class is present in the glosis_cm (common) ontology!

- were the two classes marked with "missing prefix" in the ontology by mistake? They are no longer there, but I don't think you have added any new class with the recent PR, is that right?


I have also updated the excel spreadsheet with a status to all of the additions. 

Best regards,
Bogusz

ktoddbrown commented:
At the risk of muddying the water here, you might want to consider the following about bulk density.

BulkDensityFineEarth and BulkDensityWholeSoil both refer to different sample treatments where as BulkDensityMineral and BulkDensityPeat (I think might want BulkDensityOrganic here instead?? peat is a very specific soil type where as BulkDensityOrganic is more general) refer to specific locations in the soil profile that the sample was taken from.

I see more datasets identifying the BulkDensity values by method (and it's more relevant for data harmonizations) as opposed to by location. I would consider removing BulkDensityMineral and BulkDensityPeat as variable names.

ldesousa commented:
@Montanaz0r Thank you for the review. I will be addressing these issues in the next few days.

@ktoddbrown Your comments are very welcome, we certainly need more domain expertise in this exercise. The properties `BulkDensityFineEarth` and `BulkDensityWholeSoil` are assessed with wet chemistry methods, whereas `BulkDensityMineral` and `BulkDensityPeat` are descriptive properties derived from the FAO guidelines. Good point on peat versus organic, I will verify this with the source documents.

ktoddbrown commented:
@ldesousa Again, feel free to throw references at me, but this doesn't make a whole lot of sense. Are you saying that the  `BulkDensityFineEarth` and `BulkDensityWholeSoil` are measured properties (methods generally involve extracting a fixed volume, processing the sample (sieved vs non-sieved, possibly drying) and then taking the ratio between the weight and volume). And the `BulkDensityMineral` vs `BulkDensityPeat` are given threshold values proscribed by FAO guidelines?

KathiSchleidt commented:
Hi all,
to me it looks like the GloSIS ObservableProperties are closer to what's been done in [RDA I-ADOPT](https://www.rd-alliance.org/group/interoperable-descriptions-observable-property-terminology-wg-i-adopt-wg/wiki/i-adopt-0) (still in the process of being finalized, but aligned with O&M V3)
In a nutshell, this work incorporates further concepts with the ObservableProperty concept (whereby it has now been renamed "Variable" to avoid conceptual overlaps with existing ObservableProperty concepts from O&M and SOSA), foremost:

- ObjectOfInterest: in chemical measurements, the substance being investigated (Property itself is only concentration)
- Matrix: think this may be the place for concepts such as Peat
- ContextObject: (basically the class of the FoI)

I recently did an overview [presentation ](https://www.dropbox.com/s/g61c1z2bbgxlymy/O%26M%20V3%20to%20STA.pptx?dl=0) for some OGC groups (Focus the O&M V3 update, but as of slide 15 a very basic overview of the I-ADOPT work). If you want more info, I can get you on contact with Barbara Magagna who co-chairs the I-ADOPT group

ldesousa commented:
> @ldesousa Again, feel free to throw references at me, but this doesn't make a whole lot of sense. Are you saying that the `BulkDensityFineEarth` and `BulkDensityWholeSoil` are measured properties (methods generally involve extracting a fixed volume, processing the sample (sieved vs non-sieved, possibly drying) and then taking the ratio between the weight and volume). And the `BulkDensityMineral` vs `BulkDensityPeat` are given threshold values proscribed by FAO guidelines?

Hi again Kathe. These four properties are already in the ontology, this pull request in "only" extending it with Procedures. The `BulkDensityMineral` and `BulkDensityPeat` properties make direct reference to FAO's tables:

https://github.com/rapw3k/glosis/blob/85e58f98bf2a15518ab7758c8963bab2e94dea15/glosis_layer_horizon_v2.0.0.ttl#L618

https://github.com/rapw3k/glosis/blob/85e58f98bf2a15518ab7758c8963bab2e94dea15/glosis_layer_horizon_v2.0.0.ttl#L191

These are the corresponding code-lists:

https://github.com/rapw3k/glosis/blob/85e58f98bf2a15518ab7758c8963bab2e94dea15/glosis_cl_v1.0.0.ttl#L4626

https://github.com/rapw3k/glosis/blob/85e58f98bf2a15518ab7758c8963bab2e94dea15/glosis_cl_v1.0.0.ttl#L6668

Of course none of this is set in stone, but it can be traced back to the source.

ldesousa commented:
> Hi all,
> to me it looks like the GloSIS ObservableProperties are closer to what's been done in [RDA I-ADOPT](https://www.rd-alliance.org/group/interoperable-descriptions-observable-property-terminology-wg-i-adopt-wg/wiki/i-adopt-0) (still in the process of being finalized, but aligned with O&M V3)
> In a nutshell, this work incorporates further concepts with the ObservableProperty concept (whereby it has now been renamed "Variable" to avoid conceptual overlaps with existing ObservableProperty concepts from O&M and SOSA), foremost:
> 
> * ObjectOfInterest: in chemical measurements, the substance being investigated (Property itself is only concentration)
> * Matrix: think this may be the place for concepts such as Peat
> * ContextObject: (basically the class of the FoI)
> 
> I recently did an overview [presentation ](https://www.dropbox.com/s/g61c1z2bbgxlymy/O%26M%20V3%20to%20STA.pptx?dl=0) for some OGC groups (Focus the O&M V3 update, but as of slide 15 a very basic overview of the I-ADOPT work). If you want more info, I can get you on contact with Barbara Magagna who co-chairs the I-ADOPT group

Thank you for the reference Kathi. For now I would like to complete this pull request with SOSA, I will then have a closer look at I-ADOPT. And of course would like to hear @Montanaz0r and @rapw3k on this. It might be useful to open a specific issue/discussion on this topic.

Montanaz0r commented:
> Ok I see that Mineral/Peat are essentially categorical descriptions. I would be interested in talking more about how to restructure these (BulkDensityXX), at a minimum the naming conventions are incredibly confusing as a domain scientist.

Thank you all for your engagement in this discussion. 

Yes, initially we had four BulkDensity related classes. Two of them **BulkDensityMineral** and **BulkDensityPeat** are pointing to code lists that you can treat as categories, and those are not being changed with the current PR. 
The other two (**BulkDensityWholeSoil** and **BulkDensityFineEarth** are resulting in numerical values and both are restricted to the same **unit: KiloGM-PER-DeciM3**.  Even though Luis initially added the fifth class that he name BulkDensity and extended it with a procedure, we discussed that this is possibly overlapping the two classes we already had in the ontology, namely: **BulkDensityWholeSoil** and **BulkDensityFineEarth**. So currently we are keeping all four classes that were initially in the ontology, and we are simply extending the two of them that are resulting in numerical values by adding appropriate procedure.

Now the question becomes, how much confusion the distinction between two categorical classes makes and if we can somehow unify them, as they seem to contain a similar set of values in the codelist. At least the notation is identical and the label in the case of **BulkDensityPeatValue** relates to a verbal representation of g/cm3 as per the example below

https://github.com/rapw3k/glosis/blob/85e58f98bf2a15518ab7758c8963bab2e94dea15/glosis_cl_v1.0.0.ttl#L6686-L6690

whereas **BulkDensityMineralValue** label is just a literal description

https://github.com/rapw3k/glosis/blob/85e58f98bf2a15518ab7758c8963bab2e94dea15/glosis_cl_v1.0.0.ttl#L4644-L4649

Thank you @KathiSchleidt for sharing the presentation, I will take a closer look and discuss this with @rapw3k before our meeting.

KathiSchleidt commented:
A short clarification here, the "Bulk Density" group of Observations was defined based on the requirements detailed in the FAO Guidelines on Soil Description, starting page 50. You'll find this pattern with most of the container-classes grouping Observations as this document was the main source for the data modelling work, no further guidance provided. Also - these container classes were created to ease use with the UML model, not to create deeper semantics, so utilized in this manner with care!

ldesousa commented:
@Montanaz0r I went through the remaining issues:

- **Bulk Density**: this is explained in the discussion above, the four properties derived from the UML remain as they were. It will be necessary to follow up with @ktoddbrown and other soil scientists. 

- `glosis_lh:coarseFragments`: I think it is fine to leave the two properties for now and eventually delete one later.

- `glosis_lh:cecSoil`: this property was rebaptised to `glosis_lh:CationExchangeCapacitycSoil`, to be in line with other property names and also facilitate applying the Camel case (sorry, should have explained this before). 

- `aluminum` and `aluminium`: European and American English mixed up in the spreadsheet. Now it is all European English.

- `glosis_cm:OrganicMatterClass`: this is a descriptive property, different from `glosis_lh:organicMatter`.

- **missing prefix** - there was a bug in the code, creating properties with empty names, they should have never appeared.

Apart from all this, I still need to archive the transformation programmes in the `utils` folder. But you could merge already if you wish.
